pkg/apiserver/webhook/secret_webhook.go

```diff
@@ -0,0 +1,167 @@
+	options := &controllerclient.ListOptions{Limit: 1000}
+	err := v.Client.List(context.TODO(), cpaList, options)
+	if err != nil {
+		return fmt.Errorf("%s, err:%v", errorMsg00001, err), nil
+	}
+
+	for _, cpa := range cpaList.Items {
+		v.Log.V(1).Info("Checking CloudProvideAccount", "Name", cpa.Name,
+			"Namespace", cpa.Namespace)
+		if cpa.Spec.AWSConfig != nil {
+			if cpa.Spec.AWSConfig.SecretRef.Name == s.Name &&
+				cpa.Spec.AWSConfig.SecretRef.Namespace == s.Namespace {
+				return nil, &cpa
+			}
+		}
+
+		if cpa.Spec.AzureConfig != nil {
+			if cpa.Spec.AzureConfig.SecretRef.Name == s.Name &&
+				cpa.Spec.AzureConfig.SecretRef.Namespace == s.Namespace {
+				return nil, &cpa
+			}
+		}
+	}
+	return nil, nil
+}
+
+func (v *SecretValidator) validateCreate(req admission.Request) admission.Response {
+	return admission.Allowed("")
+}
+
+func (v *SecretValidator) validateUpdate(req admission.Request) admission.Response {
+	newSecret := &corev1.Secret{}
+	err := v.decoder.Decode(req, newSecret)
+	if err != nil {
+		v.Log.Error(err, "Failed to decode Secret", "SecretValidator", req.Name)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	oldSecret := &corev1.Secret{}
+	if req.OldObject.Raw != nil {
+		if err := json.Unmarshal(req.OldObject.Raw, &oldSecret); err != nil {
+			v.Log.Error(err, "Failed to decode old Secret", "SecretValidator", req.Name)
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+	}
+
+	err, cpa := v.getCPABySecret(oldSecret)
+	if err != nil && strings.Contains(err.Error(), errorMsg00001) {
+		return admission.Denied(err.Error())
+	}
+	if cpa != nil {
+		var key string
+		if cpa.Spec.AWSConfig != nil {
+			key = cpa.Spec.AWSConfig.SecretRef.Key
+		} else if cpa.Spec.AzureConfig != nil {
+			key = cpa.Spec.AzureConfig.SecretRef.Key
+		}
+		if ok := v.allowSecretUpdate(newSecret, oldSecret, key); !ok {
+			v.Log.Info("Secret resource data cannot be modified", "Secret", oldSecret.Name,
+				"CloudProviderAccount", cpa.Name)
+			return admission.Denied(fmt.Sprintf("data %v cannot be changed for Secret %v used by "+
+				"CloudProviderAccount %s", cpa.Name, oldSecret.Name, cpa.Name))
+		}
+	}
+	return admission.Allowed("")
+}
+
+func (v *SecretValidator) validateDelete(req admission.Request) admission.Response {
+	oldSecret := &corev1.Secret{}
+	if req.OldObject.Raw != nil {
+		if err := json.Unmarshal(req.OldObject.Raw, &oldSecret); err != nil {
+			v.Log.Error(err, "Failed to decode old Secret", "SecretValidator", req.Name)
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+	}
+	err, cpa := v.getCPABySecret(oldSecret)
+	if err != nil && strings.Contains(err.Error(), errorMsg00001) {
+		return admission.Denied(err.Error())
+	}
+	if cpa != nil {
+		v.Log.Info("Secret resource cannot be deleted", "Secret", oldSecret.Name,
+			"CloudProviderAccount", cpa.Name)
+		return admission.Denied(fmt.Sprintf("cannot delete Secret %s used by CloudProviderAccount %s",
```

Comment:
Line 162, 163  convey the same info. Can you remove at time 162

Reply:
If we skip L162, then in production case, we wont have any information about the denying the request.

Reply:
Discussed offline. Decided to change the log message to ERROR level and rephrase the error.